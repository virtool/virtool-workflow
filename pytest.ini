[pytest]
addopts = --disable-pytest-warnings --docker-compose-remove-volumes
