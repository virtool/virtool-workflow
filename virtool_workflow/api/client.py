import aiohttp
from pyfixtures import fixture
from functools import wraps


@fixture
async def http():
    """:class:`Aiohttp.ClientSession` instance to be used for workflows."""
    async with aiohttp.ClientSession(auto_decompress=False) as session:
        yield JobApiHttpSession(session)


@fixture
async def authenticated_http(job_id, key, http):
    """:class:`Aiohttp.ClientSession` instance which includes authentication headers for the jobs API."""
    http.auth = aiohttp.BasicAuth(login=f"job-{job_id}", password=key)
    return http


class JobApiHttpSession:
    """Wraps :class:`aiohttp.ClientSession` and adds authentication for the jobs API."""

    def __init__(self, client: aiohttp.ClientSession, auth: aiohttp.BasicAuth = None):
        self.client = client
        self.auth = auth

        self.delete = self._wrap_with_auth(self.client.delete)
        self.get = self._wrap_with_auth(self.client.get)
        self.patch = self._wrap_with_auth(self.client.patch)
        self.post = self._wrap_with_auth(self.client.post)
        self.put = self._wrap_with_auth(self.client.put)

    def _wrap_with_auth(self, method):
        @wraps(method)
        def _method_with_auth(*args, noauth=False, **kwargs):
            if "auth" not in kwargs and self.auth is not None and not noauth:
                kwargs["auth"] = self.auth

            return method(*args, **kwargs)

        return _method_with_auth
