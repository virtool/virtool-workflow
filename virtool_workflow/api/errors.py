from contextlib import asynccontextmanager
from logging import getLogger
from typing import Dict, Type, List

logger = getLogger("api")


class JobAlreadyAcquired(Exception):
    def __init__(self, job_id: str):
        super(JobAlreadyAcquired, self).__init__(
            f"Job {job_id} is has already been acquired."
        )


class JobsAPIServerError(Exception):
    ...


class InsufficientJobRights(Exception):
    ...


class NotFound(KeyError):
    ...


class AlreadyFinalized(Exception):
    ...


@asynccontextmanager
async def raising_errors_by_status_code(
    response,
    accept: List[int] = None,
    status_codes_to_exceptions: Dict[int, Type[Exception]] = None,
):
    """
    Raise exceptions based on the result status code.

    If the status code is between 200 and 299 then this context manager will have no effect, other than
    getting the JSON body of the response.

    :param response: The aiohttp response object.
    :param accept: A list of status codes to consider successful. Defaults to 200-299.
    :param status_codes_to_exceptions: A dict associating status codes to exceptions that should be raised.
    :return: The response json as a dict, if it is available.
    """
    if status_codes_to_exceptions is None:
        status_codes_to_exceptions = {
            404: NotFound,
            409: AlreadyFinalized,
            403: InsufficientJobRights,
            500: JobsAPIServerError,
        }

    if accept is None:
        accept = list(range(200, 299))

    response_json = None

    if response.content_type == "application/json":
        try:
            response_json = await response.json()
        except UnicodeDecodeError:
            pass

    if response.status in accept:
        yield response_json
        return

    logger.info(await response.text())

    if response.status in status_codes_to_exceptions:
        if response_json:
            response_message = (
                response_json["message"]
                if "message" in response_json
                else str(response_json)
            )
        else:
            try:
                response_message = await response.text()
            except UnicodeDecodeError:
                response_message = "Could not decode response message"
        raise status_codes_to_exceptions[response.status](response_message)
    else:
        raise ValueError(
            f"Status code {response.status} not handled for response\n {response}"
        )
