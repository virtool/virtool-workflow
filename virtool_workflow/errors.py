class IllegalJobArguments(ValueError):
    """The `job.args` dict is in an illegal state."""

    ...


class MissingJobArgument(ValueError):
    """The `job.args` dict is missing a required key for some funcionality."""

    ...
