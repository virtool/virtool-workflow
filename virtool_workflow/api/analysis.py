"""
A data provider for Virtool analysis based on HTTP communication with the Virtool job API.

"""
import logging
from pathlib import Path
from typing import Dict, Any, Tuple, List

import aiofiles
import aiohttp
import dateutil.parser

from virtool_workflow.api.errors import raising_errors_by_status_code
from virtool_workflow.api.utils import upload_file_via_put
from virtool_workflow.data_model.analysis import Analysis
from virtool_workflow.data_model.files import VirtoolFile, VirtoolFileFormat

logger = logging.getLogger(__name__)


def get_analysis_files_from_response_json(json) -> List[VirtoolFile]:
    """
    Generate a list of :class:`.VirtoolFile` objects from a JSON response representing a Virtool analysis.

    :param json: the analysis JSON data
    :return: the list of :class:`.VirtoolFile` objects for the analysis

    """
    try:
        return [
            VirtoolFile(
                id=f["id"],
                name=f["name"],
                name_on_disk=f["name_on_disk"],
                size=f["size"],
                uploaded_at=dateutil.parser.isoparse(f["uploaded_at"]),
                format=f["format"]
            )
            for f in json["files"]
        ]
    except KeyError as error:
        if error.args[0] == "files":
            return []
        raise


async def get_analysis_by_id(analysis_id: str, http: aiohttp.ClientSession, jobs_api_url: str) -> Analysis:
    """
    Get the analysis by id via the jobs API.

    :param analysis_id: The id of the analysis.
    :param http: A :class:`aiohttp.ClientSession` instance.
    :param jobs_api_url: The url for the jobs API server.

    :return: A :class:`virtool_workflow.data_model.analysis.Analysis` instance.

    :raise InsufficientJobRights: When the current job does not have sufficient rights to access the analysis.
    :raise NotFound: When the given :obj:`analysis_id` does not correspond to an existing analysis (HTTP 404).
    :raise KeyError: When the analysis data received from the API is missing a required key.

    """
    async with http.get(f"{jobs_api_url}/analyses/{analysis_id}") as response:
        async with raising_errors_by_status_code(response) as response_json:
            return Analysis(
                id=response_json["id"],
                files=get_analysis_files_from_response_json(response_json),
                ready=response_json["ready"] if "ready" in response_json else False
            )


class AnalysisProvider:
    """
    Use the Virtool Jobs API to perform operations on the current analysis.

    :param analysis_id: The ID of the current analysis as found in the job args.
    :param http: A :class:`aiohttp.ClientSession` instance to be used when making requests.
    :param jobs_api_url: The url to the Jobs API.

    """

    def __init__(self,
                 analysis_id: str,
                 http: aiohttp.ClientSession,
                 jobs_api_url: str):
        self.id = analysis_id
        self.http = http
        self.api_url = jobs_api_url

    async def get(self) -> Analysis:
        """
        Fetch the current analysis.

        """
        return await get_analysis_by_id(self.id, self.http, self.api_url)

    async def upload(self, path: Path, format: VirtoolFileFormat):
        """
        Upload a file in the workflow environment that should be associated with the current analysis.

        :param path: the path to the file to upload
        :param format: the format of the file

        """
        return await upload_file_via_put(
            self.http,
            f"{self.api_url}/analyses/{self.id}/files",
            path,
            format
        )

    async def download(self, file_id: str, target_path: Path) -> Path:
        """
        Download a file associated with the current analysis.

        :param file_id: The ID of the file.
        :param target_path: The path which the file data should be stored under.

        :return: A path to the downloaded file. It will be the `target_path` if one was given.

        :raise NotFound: When either the file or the analysis does not exist (404 status code).

        """
        async with self.http.get(f"{self.api_url}/analyses/{self.id}/files/{file_id}") as response:
            async with raising_errors_by_status_code(response):
                async with aiofiles.open(target_path, "wb") as f:
                    await f.write(await response.read())

        return target_path

    async def upload_result(self, result: Dict[str, Any]) -> Tuple[Analysis, dict]:
        """
        Upload the results dict for the analysis.

        :param result: The results dict.

        :raise InsufficientJobRights: When the current job does not have sufficient rights to modify the analysis.
        :raise NotFound: When there is no analysis with the ID :obj:`.id`.
        :raise AlreadyFinalized: When there is already a result for the analysis.

        """
        async with self.http.patch(f"{self.api_url}/analyses/{self.id}", json={
            "results": result
        }) as response:
            async with raising_errors_by_status_code(response) as analysis_json:
                return Analysis(
                    analysis_json["id"],
                    get_analysis_files_from_response_json(analysis_json)
                ), analysis_json["results"]

    async def delete(self):
        """
        Delete the analysis. This method should be called if the workflow code fails before a result is uploaded.

        :raise InsufficientJobRights: When the current job does not have the appropriate permissions required
            to delete the analysis.
        :raise NotFound: When the analysis has already been deleted.
        :raise AlreadyFinalized: When the analysis has a result uploaded and is viewable in Virtool.
            Jobs are not permitted to delete the analysis after this point.
        """
        async with self.http.delete(f"{self.api_url}/analyses/{self.id}") as response:
            async with raising_errors_by_status_code(response):
                return

    def __await__(self) -> Analysis:
        return self.get().__await__()
