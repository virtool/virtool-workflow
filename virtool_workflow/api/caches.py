import asyncio
import shutil
from pathlib import Path

import aiohttp
from virtool_workflow.api.errors import (JobsAPIServerError, NotFound,
                                         raising_errors_by_status_code)
from virtool_workflow.api.utils import (read_file_from_response,
                                        upload_file_via_put)
from virtool_workflow.caching.caches import (Cache, Caches,
                                             CacheWriter, ReadsCache)
from virtool_workflow.data_model.files import VirtoolFileFormat
from virtool_workflow.execution.run_in_executor import FunctionExecutor


class CacheAlreadyOpen(Exception):
    pass


class RemoteReadsCacheWriter(CacheWriter[ReadsCache]):

    def __init__(self, key, path, sample_id, http, jobs_api_url, run_in_executor):
        super(RemoteReadsCacheWriter, self).__init__(key, path)
        self.http = http
        self.sample_id = sample_id
        self.url = f"{jobs_api_url}/samples/{sample_id}/caches"
        self.run_in_executor = run_in_executor

    async def open(self):
        """
        Create a new cache if one does not exist.

        :raise CacheAlreadyOpen: When there is an existing cache.
        :raise NotFound: When `.sample_id` does not identify a sample.
        :raise JobsAPIServerError: When there is any unexpected status code in the server's response.
        """
        response = await self.http.post(self.url, json={
            "key": self.key
        })

        if response.status == 201:
            return

        if response.status == 409:
            raise CacheAlreadyOpen(self.key)
        elif response.status == 404:
            raise NotFound(self.sample_id)
        else:
            raise JobsAPIServerError(response.status)

    async def upload(self, path: Path, format_: VirtoolFileFormat = "fastq"):
        """Upload a file to the cache."""
        await self.run_in_executor(shutil.copyfile, path, self.path / path.name)

        if path.name in ("reads_1.fq.gz", "reads_2.fq.gz"):
            return await upload_file_via_put(self.http,
                                             f"{self.url}/{self.key}/reads/{path.name}",
                                             path,
                                             params={})

        return await upload_file_via_put(self.http, f"{self.url}/{self.key}/artifacts", path, params={
            "name": path.name,
            "type": format_,
        })

    async def close(self):
        """Finalize the cache."""
        await super(RemoteReadsCacheWriter, self).close()

        response = await self.http.patch(f"{self.url}/{self.key}", json={
            "quality": self.cache.quality
        })

        if response.status != 200:
            raise JobsAPIServerError(response.status)

    async def delete(self):
        """Delete the cache."""
        response = await self.http.delete(f"{self.url}/{self.key}")
        async with raising_errors_by_status_code(response, accept=[204]):
            pass


class RemoteReadCaches(Caches[ReadsCache]):
    def __init__(
            self,
            sample_id: str,
            paired: bool,
            path: Path,
            http: aiohttp.ClientSession,
            jobs_api_url: str,
            run_in_executor: FunctionExecutor,
            poll_rate: int = 5,
    ):
        self.sample_id = sample_id
        self.paired = paired
        self.path = path
        self.http = http
        self.url = f"{jobs_api_url}/samples/{self.sample_id}/caches"
        self.jobs_api_url = jobs_api_url
        self.run_in_executor = run_in_executor
        self.poll_rate = poll_rate

    async def get(self, key: str) -> ReadsCache:
        """
        Get the cache with the given key.

        If a cache exists, but is not ready, this method will not return until the
        cache is ready. This occurs when multiple instances of a workflow are attempting
        to create a cache with the same key.
        """
        response = await self.http.get(f"{self.url}/{key}")

        if response.status == 404:
            raise KeyError(key)

        if response.status != 200:
            raise JobsAPIServerError(response.status)

        cache = await response.json()

        if cache["ready"] is not True:
            # Continually request cache until it is ready.
            await asyncio.sleep(self.poll_rate)
            return await self.get(key)

        cache_path = self.path / key
        cache_path.mkdir()

        files = ("reads_1.fq.gz", "reads_2.fq.gz") if self.paired else (
            "reads_1.fq.gz",)

        for filename in files:
            response = await self.http.get(f"{self.url}/{key}/reads/{filename}")
            await read_file_from_response(response, cache_path)

        return ReadsCache(
            key=key,
            path=cache_path,
            quality=cache["quality"],
        )

    def create(self, key: str) -> CacheWriter[Cache]:
        """Create a new cache writer."""
        cache_path = self.path / key
        cache_path.mkdir()
        return RemoteReadsCacheWriter(key, cache_path, self.sample_id,
                                      self.http, self.jobs_api_url, self.run_in_executor)
