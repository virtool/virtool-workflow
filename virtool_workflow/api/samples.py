import logging
import pprint
from pathlib import Path
from typing import Dict, Any

import aiohttp

from virtool_workflow.analysis.utils import ReadPaths, make_read_paths
from virtool_workflow.api.errors import (raising_errors_by_status_code,
                                         AlreadyFinalized,
                                         JobsAPIServerError)
from virtool_workflow.api.utils import (upload_file_via_put,
                                        read_file_from_response)
from virtool_workflow.data_model import Sample
from virtool_workflow.data_model.files import VirtoolFileFormat, VirtoolFile

logger = logging.getLogger(__name__)


async def _make_sample_from_response(response) -> Sample:
    async with raising_errors_by_status_code(response) as sample_json:
        logger.debug(pprint.pformat(sample_json))
        return Sample(
            id=sample_json["id"],
            name=sample_json["name"],
            host=sample_json["host"],
            isolate=sample_json["isolate"],
            locale=sample_json["locale"],
            library_type=sample_json["library_type"],
            paired=sample_json["paired"],
            quality=sample_json["quality"],
            nuvs=sample_json["nuvs"],
            pathoscope=sample_json["pathoscope"],
            files=sample_json["files"] if "files" in sample_json else [],
        )


class SampleProvider:

    def __init__(self,
                 sample_id: str,
                 http: aiohttp.ClientSession,
                 jobs_api_url: str):
        self.id = sample_id
        self.http = http
        self.url = f"{jobs_api_url}/samples/{sample_id}"

    async def get(self) -> Sample:
        async with self.http.get(self.url) as response:
            logger.info("Fetched sample document")
            return await _make_sample_from_response(response)

    async def finalize(self, quality: Dict[str, Any]) -> Sample:
        async with self.http.patch(self.url, json={"quality": quality}) as response:
            return await _make_sample_from_response(response)

    async def delete(self):
        async with self.http.delete(self.url) as response:
            async with raising_errors_by_status_code(response, accept=[204], status_codes_to_exceptions={
                400: AlreadyFinalized,
                500: JobsAPIServerError
            }):
                pass

    async def upload(self, path: Path, format: VirtoolFileFormat = "fastq") -> VirtoolFile:
        if path.name in ("reads_1.fq.gz", "reads_2.fq.gz"):
            return await upload_file_via_put(self.http, f"{self.url}/reads/{path.name}", path, params={})

        return await upload_file_via_put(self.http, f"{self.url}/artifacts", path, params={
            "name": path.name,
            "type": format
        })

    async def download_reads(self, target_path: Path, paired: bool = None) -> ReadPaths:
        if paired is None:
            sample = await self.get()
            paired = sample.paired

        async with self.http.get(f"{self.url}/reads/reads_1.fq.gz") as response:
            await read_file_from_response(response, target_path / "reads_1.fq.gz")

        if paired:
            async with self.http.get(f"{self.url}/reads/reads_2.fq.gz") as response:
                await read_file_from_response(response, target_path / "reads_2.fq.gz")

        return make_read_paths(target_path, paired)

    async def download_artifact(self, filename: str, target_path: Path):
        async with self.http.get(f"{self.url}/artifacts/{filename}") as response:
            await read_file_from_response(response, target_path / filename)
