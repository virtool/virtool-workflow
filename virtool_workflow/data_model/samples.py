from pathlib import Path
from typing import Optional

from virtool_core.models.samples import Sample


class WFSample(Sample):
    """A Virtool Sample."""

    reads_path: Optional[Path] = None
    read_paths: Optional[Path] = None

    @property
    def min_length(self) -> Optional[int]:
        """
        The minimum observed read length in the sample sequencing data.

        Returns ``None`` if the sample is still being created and no quality data is available.

        """
        return self.quality.length[0] if self.quality else None

    @property
    def max_length(self) -> Optional[int]:
        """
        The maximum observed read length in the sample sequencing data.

        Returns ``None`` if the sample is still being created and no quality data is available.

        """
        return self.quality.length[1] if self.quality else None
