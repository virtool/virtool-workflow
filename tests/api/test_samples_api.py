import gzip
from pathlib import Path

import pytest

from tests.api.mocks.mock_sample_routes import TEST_SAMPLE_ID
from virtool_workflow.api.errors import AlreadyFinalized
from virtool_workflow.api.samples import SampleProvider

QUALITY = {
    "bases": [
        [30, 31, 31, 31, 31, 31],
        [30, 31, 31, 31, 31, 32],
    ],
    "composition": [
        [32, 14, 4, 40],
        [22, 12, 34, 25],
    ],
    "count": 998822,
    "encoding": "Sanger / Illumina 1.8",
    "gc": 43,
    "length": [0, 52],
    "sequences": [
        0,
        0,
        3512,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
    ],
}


@pytest.fixture
def sample_api(http, jobs_api_connection_string: str):
    return SampleProvider(TEST_SAMPLE_ID, http, jobs_api_connection_string)


async def test_get(sample_api, snapshot):
    assert await sample_api.get() == snapshot


async def test_finalize(sample_api, snapshot):
    sample = await sample_api.finalize(QUALITY)

    assert sample.quality == QUALITY
    assert sample == snapshot


async def test_delete(sample_api):
    await sample_api.delete()


async def test_delete_after_finalize(sample_api):
    sample_api.id = "finalized"

    with pytest.raises(AlreadyFinalized):
        await sample_api.delete()


async def test_upload(sample_api, tmpdir):
    tmpdir = Path(tmpdir)

    reads_1 = tmpdir / "reads_1.fq.gz"
    artifact = tmpdir / "artifact.json"

    reads_1.touch()
    artifact.touch()

    await sample_api.upload(reads_1)
    await sample_api.upload(artifact)


async def test_download(sample_api, tmpdir, file_regression):
    tmpdir = Path(tmpdir)

    read_files = await sample_api.download_reads(tmpdir, paired=True)

    for read_file in read_files:
        with gzip.open(read_file) as f:
            file_regression.check(f.read(), basename=read_file.name, binary=True)

    assert (
        read_files == read_files == (tmpdir / "reads_1.fq.gz", tmpdir / "reads_2.fq.gz")
    )

    await sample_api.download_artifact("artifact.json", tmpdir)

    assert (tmpdir / "artifact.json").exists()
