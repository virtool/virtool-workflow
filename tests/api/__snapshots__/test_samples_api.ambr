# name: test_finalize
  dict({
    'all_read': True,
    'all_write': True,
    'artifacts': list([
    ]),
    'caches': list([
    ]),
    'created_at': datetime.datetime(2021, 1, 11, 18, 24, 33, 149000, tzinfo=datetime.timezone.utc),
    'format': 'fastq',
    'group': 'none',
    'group_read': True,
    'group_write': True,
    'hold': True,
    'host': '',
    'id': 'hl5v0i0y',
    'is_legacy': False,
    'isolate': '',
    'labels': list([
    ]),
    'library_type': <LibraryType.normal: 'normal'>,
    'locale': '',
    'name': '20PRT_3_A2_10Mil',
    'notes': '',
    'nuvs': False,
    'paired': False,
    'pathoscope': True,
    'quality': dict({
      'bases': list([
        list([
          30,
          31,
          31,
          31,
          31,
          31,
        ]),
        list([
          30,
          31,
          31,
          31,
          31,
          32,
        ]),
      ]),
      'composition': list([
        list([
          32,
          14,
          4,
          40,
        ]),
        list([
          22,
          12,
          34,
          25,
        ]),
      ]),
      'count': 998822,
      'encoding': 'Sanger / Illumina 1.8',
      'gc': 43,
      'length': list([
        0,
        52,
      ]),
      'sequences': list([
        0,
        0,
        3512,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
      ]),
    }),
    'read_paths': None,
    'reads': list([
      dict({
        'download_url': '/samples/hl5v0i0y/reads/reads_1.fq.gz',
        'id': 1129,
        'name': 'reads_1.fq.gz',
        'name_on_disk': 'reads_1.fq.gz',
        'sample': 'hl5v0i0y',
        'size': 30696061,
        'upload': None,
        'uploaded_at': datetime.datetime(2022, 9, 21, 15, 43, 58, 423000, tzinfo=datetime.timezone.utc),
      }),
    ]),
    'reads_path': None,
    'ready': True,
    'subtractions': list([
      dict({
        'id': '123',
        'name': 'Vitis vinifera + plastids',
      }),
    ]),
    'user': dict({
      'administrator': False,
      'handle': 'james',
      'id': 'abc12345',
    }),
  })
# ---
# name: test_get
  dict({
    'all_read': True,
    'all_write': True,
    'artifacts': list([
    ]),
    'caches': list([
    ]),
    'created_at': datetime.datetime(2021, 1, 11, 18, 24, 33, 149000, tzinfo=datetime.timezone.utc),
    'format': 'fastq',
    'group': 'none',
    'group_read': True,
    'group_write': True,
    'hold': True,
    'host': '',
    'id': 'hl5v0i0y',
    'is_legacy': False,
    'isolate': '',
    'labels': list([
    ]),
    'library_type': <LibraryType.normal: 'normal'>,
    'locale': '',
    'name': '20PRT_3_A2_10Mil',
    'notes': '',
    'nuvs': False,
    'paired': False,
    'pathoscope': True,
    'quality': dict({
      'bases': list([
        list([
          31,
          32,
          32,
          32,
          32,
          32,
        ]),
        list([
          31,
          32,
          32,
          32,
          32,
          32,
        ]),
        list([
          31,
          32,
          32,
          32,
          32,
          32,
        ]),
        list([
          31,
          32,
          32,
          32,
          32,
          32,
        ]),
        list([
          31,
          32,
          32,
          32,
          32,
          32,
        ]),
      ]),
      'composition': list([
        list([
          37,
          12,
          7,
          41,
        ]),
        list([
          23,
          13,
          35,
          26,
        ]),
        list([
          20,
          17,
          27,
          34,
        ]),
        list([
          23,
          24,
          25,
          26,
        ]),
        list([
          24,
          30,
          25,
          19,
        ]),
      ]),
      'count': 10000000,
      'encoding': 'Sanger / Illumina 1.9',
      'gc': 43,
      'length': list([
        35,
        76,
      ]),
      'sequences': list([
        0,
        0,
        3512,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        2,
        1,
        34,
        164,
        640,
        2044,
        5589,
        14429,
        32328,
      ]),
    }),
    'read_paths': None,
    'reads': list([
      dict({
        'download_url': '/samples/hl5v0i0y/reads/reads_1.fq.gz',
        'id': 1129,
        'name': 'reads_1.fq.gz',
        'name_on_disk': 'reads_1.fq.gz',
        'sample': 'hl5v0i0y',
        'size': 30696061,
        'upload': None,
        'uploaded_at': datetime.datetime(2022, 9, 21, 15, 43, 58, 423000, tzinfo=datetime.timezone.utc),
      }),
    ]),
    'reads_path': None,
    'ready': True,
    'subtractions': list([
      dict({
        'id': '123',
        'name': 'Vitis vinifera + plastids',
      }),
    ]),
    'user': dict({
      'administrator': False,
      'handle': 'james',
      'id': 'abc12345',
    }),
  })
# ---
